dashboard/app/views/levels/_level_group.haml

```diff
@@ -6,64 +6,86 @@
-    - last_attempt = last_attempt_entry["result"] if last_attempt_entry
+-#
+  If we are looking at the level independent of a script, then let's just
+  show all pages of the LevelGroup at once.  Otherwise we'll show the actual
+  page.
+
+- if @script_level.nil?
+  - pages = (0..data["pages"].length-1).to_a
+- else
+  - current_page = params[:puzzle_page] ? (params[:puzzle_page].to_i - 1) : 0
+  - pages = [current_page]
+
+- pages.each do |page|
+
+  - unless data["title"].blank?
+    %h2= t('heading_x_of_y', heading: data["title"], x: page+1, y: data["pages"].length)
+    %br/
+
+  - page_levels = data["pages"][page]["levels"]
+
+  - if page == 0
+    - prior_page_count = 0
   - else
-    - last_attempt = ""
-  - level_class = level.class.to_s.underscore
-
-  - if level_class == "multi"
-    -# For students (or teachers doing PD), mark all answers incorrect, so we don't give anything away.
-    - unless current_user.try(:authorized_teacher?) && !@script.try(:pd?)
-      - level.properties['answers'].each { |answer| answer['correct'] = false }
-    = render partial: 'levels/single_multi', locals: {page: page, index: index, standalone: false, level: level, last_attempt: last_attempt, tight_layout: true }
-  - elsif level_class = "text_match"
-    -# For students (or teachers doing PD), clear out all answers.
-    - unless current_user.try(:authorized_teacher?) && !@script.try(:pd?)
-      - level.properties['answers'] = nil
-    = render partial: 'levels/single_text_match', locals: {page: page, index: index, standalone: false, level: level, last_attempt: last_attempt }
+    - prior_page_count = data["pages"][0..page-1].inject(0){|sum,p| sum + p["levels"].length}
 
-= render partial: 'levels/common_audio'
+  - page_levels.each_with_index do |level_name, index|
+    - level = Level.find_by_name(level_name)
+    - if @last_attempt
+      - last_attempt_entry = JSON.parse(@last_attempt).find{|item| item["level_id"]==level.id}
+      - last_attempt = last_attempt_entry["result"] if last_attempt_entry
+    - else
+      - last_attempt = ""
+    - level_class = level.class.to_s.underscore
 
-- if page < data["pages"].length - 1
-  %a.btn.btn-large.btn-primary.nextPageButton
-    =t('next_page')
-- else
-  = render partial: 'levels/dialog', locals: {app: app, data: data}
-  %a.btn.btn-large.btn-primary.unsubmitButton{style: "display: none"}
-    =t('unsubmit')
+    - level.properties['title'] = nil
 
+    - if level_class == "multi"
+      -# For students (or teachers doing PD), mark all answers incorrect, so we don't give anything away.
+      - unless current_user.try(:authorized_teacher?) && !@script.try(:pd?)
+        - level.properties['answers'].each { |answer| answer['correct'] = false }
+      = render partial: 'levels/single_multi', locals: {page: page, index: index, number: prior_page_count + index + 1, standalone: false, level: level, last_attempt: last_attempt, tight_layout: true }
```

Comment:
ah, I see. I see a few slightly weird things here:
* calling db-query-like methods (like Level.find_by_name) in a view
* modifying the level properties in a view
* calculations that seem prone to off by one errors (ie. containing "+ 1"...)

Which kind of lead me to believe that more of this should be moved into the controller and model. For example, the LevelGroup model could construct something that would be a collection of levels decorated with the page context, so you could call something like this in the view:

```
- @pages.each do |page|
  - page.levels.each_with_index do |level, index|
    = render partial: 'levels/single_text_match', locals: {page: @page.page, index: index, number: @page.offset + index + 1, standalone: false, level: level, last_attempt: last_attempt}
  end
end
```

and in the controller you'd have something like:

```
# LevelController
@pages = LevelGroup.pages
```

```
#ScriptLevelController
@pages = [LevelGroup.pages[params[:page] - 1]]
```

model is:

```
class LevelGroup
  class LevelGroupPage
    def initialize(page_properties, page_number, offset)
      @levels = ....
      @offset = offset
      @page_number = page_number
    end
  end
  def pages
    offset = 0
    page_count = 0 
    @pages ||= properties['pages'].map do |page|
      # these are prob full of off by one errors
      offset += page.count
      page_count += 1
      LevelGroupPage.new(page, page_count, offset)
    end
  end
end
```

(that way the view does not need to check what controller it's in which is kind of backwards)